meta-dts-distro/recipes-dts/images/dts-base-image.inc

```
@@ -32,3 +34,9 @@ local_autologin () {
 }
 
 ROOTFS_POSTPROCESS_COMMAND += "local_autologin; "
+
+do_image_wic:prepend() {
+    WKS_FILE="${WORKDIR}/usb-stick-dts.wks"
+    UUID="${@'$(echo ${LAYER_COMMIT_SHA} | cut -c1-8)-$(echo ${LAYER_COMMIT_SHA} | cut -c9-12)-4$(echo ${LAYER_COMMIT_SHA} | cut -c13-15)-$(echo ${LAYER_COMMIT_SHA} | cut -c16-19)-$(echo ${LAYER_COMMIT_SHA} | cut -c20-31)'}"
+    sed -i "s/--use-uuid/--uuid $UUID/" $WKS_FILE
```

Don't we use: [usb-stick-dts.wks.in](https://github.com/Dasharo/meta-dts/blob/main/meta-dts-distro/wic/usb-stick-dts.wks.in)?
You can use variables directly e.g. https://github.com/Dasharo/meta-dts/blob/uki/meta-dts-distro/wic/usb-stick-dts.wks.in#L3 (setting LABEL)

---

The problem with this is the following:
We need to evaluate a bash command to create a uuid based on the commit SHA. I tried defining the UUIDs in this recipe and passing them to `WICVARS`
```
 ROOT_UUID="${@'$(echo ${LAYER_COMMIT_SHA} | sha256sum | cut -c1-8)-$(echo ${LAYER_COMMIT_SHA} | sha256sum | cut -c9-12)-4$(echo ${LAYER_COMMIT_SHA} | sha256sum | cut -c13-15)-$(echo ${LAYER_COMMIT_SHA} | sha256sum | cut -c16-19)-$(echo ${LAYER_COMMIT_SHA} | sha256sum | cut -c20-31)'}"
BOOT_UUID="${@'$(echo "boot-${LAYER_COMMIT_SHA}" | sha256sum | cut -c1-8)-$(echo "boot-${LAYER_COMMIT_SHA}" | sha256sum | cut -c9-12)-4$(echo "boot-${LAYER_COMMIT_SHA}" | sha256sum | cut -c13-15)-$(echo "boot-${LAYER_COMMIT_SHA}" | sha256sum | cut -c16-19)-$(echo "boot-${LAYER_COMMIT_SHA}" | sha256sum | cut -c20-31)'}"

WICVARS += "ROOT_UUID BOOT_UUID "
```

but the commands were not evaluated, so the variables were passed to wks in the form of commands. I'm not sure how else I could pass them aside from the original idea. Unless we want to set fixed UUIDs, then it becomes a simple matter.

---

Look at how every other recipe does this. Usually with python. I'm not sure if sh evaluation works:

```
https://docs.yoctoproject.org/bitbake/2.2/bitbake-user-manual/bitbake-user-manual-metadata.html#inline-python-variable-expansion
```

You could just define task and set value of this variable there: https://docs.yoctoproject.org/bitbake/2.2/bitbake-user-manual/bitbake-user-manual-metadata.html#exporting-variables-to-the-environment

You might write python task to begin with: https://docs.yoctoproject.org/bitbake/2.2/bitbake-user-manual/bitbake-user-manual-metadata.html#bitbake-style-python-functions
or just use this: https://docs.yoctoproject.org/bitbake/2.2/bitbake-user-manual/bitbake-user-manual-metadata.html#python-functions

```bitbake
BOOT_UUID="${@get_uuid(boot)}"
ROOT_UUID="${@get_uuid(boot)}"

def get_uuid(string):
    import hashlib, uuid
    return str(uuid.UUID(hashlib.md5(f"{commit}{string}".encode()).hexdigest()))
```

you need to somehow get commit in python func. Not tested so there probably are mistakes.

---

I found `do_rootfs_wicenv`: https://git.yoctoproject.org/poky/plain/meta/classes-recipe/image_types_wic.bbclass
Maybe I can append the evaluated variables to the <image>.env file there

---

If this doesn't work you might have to use: https://docs.yoctoproject.org/bitbake/2.2/bitbake-user-manual/bitbake-user-manual-metadata.html#immediate-variable-expansion

---

> Maybe I can append the evaluated variables to the .env file there

why?

> but the commands were not evaluated, so the variables were passed to wks in the form of commands

I think the problem was that you used wrong format. You can also use immediate variable expansion to make sure.
I think you tried to use bash script in python evaluation ('@'). Not only that but it's possible that it's because of `'` in `${@'`. It's possible that everything between `''` was treated as string

---

Done https://github.com/Dasharo/meta-dts/compare/6615c9317c0ff58faffbc46f06dade8bc31b7f6c..81be646bee124af129961089c12c970c1e6e60e4

---

You forgot to include commit sha. I think it should be something like:

```
BOOT_UUID="${@get_uuid('boot' + d.getVar('LAYER_COMMIT_SHA'))}"
ROOT_UUID="${@get_uuid('root' + d.getVar('LAYER_COMMIT_SHA'))}"
```

Check if it works (and check if `LAYER_COMMIT_SHA` is evaluated correctly)

---

Working example:

* [zarhus-base-image.inc#L12](https://github.com/zarhus/meta-zarhus/blob/ca0d27ec46afe58bfeee48e314e56bb6f6558a7c/meta-zarhus-distro/recipes-zarhus/images/zarhus-base-image.inc#L12)
* [zarhus-uki.inc#L11](https://github.com/zarhus/meta-zarhus/blob/ca0d27ec46afe58bfeee48e314e56bb6f6558a7c/meta-zarhus-bsp/meta-zarhus-bsp-x86_64/recipes-zarhus/images/zarhus-uki.inc#L11)
* [zarhus-uki.wks#L3](https://github.com/zarhus/meta-zarhus/blob/ca0d27ec46afe58bfeee48e314e56bb6f6558a7c/meta-zarhus-bsp/meta-zarhus-bsp-x86_64/wic/zarhus-uki.wks#L3) - notice `--ptable gpt` in previous line, without it wic generates disk image with MBR which uses different UID format and will silently ignore `--uuid`, at least when it's in wrong format.